packages/opentelemetry-sdk-trace-base/src/Span.ts

```diff
@@ -48,17 +48,23 @@ import {
   SEMATTRS_EXCEPTION_STACKTRACE,
   SEMATTRS_EXCEPTION_TYPE,
 } from '@opentelemetry/semantic-conventions';
-import { ExceptionEventName } from './enums';
 import { ReadableSpan } from './export/ReadableSpan';
+import { ExceptionEventName } from './enums';
 import { SpanProcessor } from './SpanProcessor';
 import { TimedEvent } from './TimedEvent';
 import { Tracer } from './Tracer';
 import { SpanLimits } from './types';
 
+/**
+ * This type provides the properties of @link{ReadableSpan} at the same time
+ * of the Span API
+ */
+export type Span = APISpan & ReadableSpan;
```

Comment:
Out of curiosity. Wouldn't be having this issue already since the class implementation was

```js
export class Span implements APISpan, ReadableSpan { ... }
```
?

> Since the SDK extension interface ends up in the same role as an API implementor, its public interface breaks based on which API version is installed

this will happen at compile time and also a runtime right? So if a user has a newer API version with an old SDK the Span returned by the tracer will lack some properties or methods (the ones added in the API). Does that mean we need to have a SDK which is backwards compat but also future proof?

>  I think to be able to use the interface like this, we'd have to duplicate all current methods from the APISpan here in this package, and then add any new ones that are added to the APISpan type as optional properties.

why optional properties? if they are defined in the API the implementor should have them right?

Reply:
>> Since the SDK extension interface ends up in the same role as an API implementor, its public interface breaks based on which API version is installed
> 
> this will happen at compile time and also a runtime right? So if a user has a newer API version with an old SDK the Span returned by the tracer will lack some properties or methods (the ones added in the API). 

Yes this would happen. It's one of the ways that the API is very If the user registers an old SDK with a newer API version - they will be violating `peerDependency` ranges of the SDK. 

> Does that mean we need to have a SDK which is backwards compat but also future proof?

I think it's kind of out of scope when talking about the API and the SDK interaction - it's a design level challenge that we will not tackle in SDK 2.0 - Nev did a write up over at #4269 a while ago. I just realized that my comment from earlier is void. Here's how I came to that conclusion:

I was thinking of possible interactions within the SDK extensions that implement interfaces of older SDK versions. 

Let's say one implements a `SpanProcessor` and offers that as a `npm` package. I will depend on `@opentelemetry/sdk-trace-base@2.0.0`. The interface for `@opentelemetry/sdk-trace-node@2.1.0` will require a `APISpan` with an additional function, so I cannot assign my `SpanProcessor` to a `TracerProvider` from `@opentelemetry/sdk-trace-node@2.1.0` - the old definition of `Span` is, after all, missing that property that was added in `@opentelemetry/sdk-trace-node@2.1.0` even though I don't even need it. To get to this point, my the direct dependency on `@opentelemetry/sdk-trace-base@2.0.0` would violate the `peerDependency` range of the `@opentelemetry/sdk-trace-node@2.1.0` - so it's the same safeguard at play as is with the API and this should never happen unless:
- end-user ignores the `peerDependency` range
- my package that offers the `SpanProcessor` does not specify the `peerDependency` on the API correctly and therefore enables the end-user to ignore the `peerDependency` range.

TL;DR: keeping this PR as-is is most likely fine given that the constraints that we're working with don't change before the next major version (which I don't expect to happen). My brain just was not working correctly. :brain:

Reply:
thanks for the thorough explanation @pichlermarc :)

I think either option is fine so I don't mind to have an explicit interface. Shall I change it?

Reply:
Let's leave it as-is. :slightly_smiling_face: